### ISSUE TITLE
OS X: Minimization into dock icon makes the window disappear completely

### ISSUE THREAD

**Issue by balayanr** [pyzo/pyzo#573]
I have minimization into icon turned on on my Mac and when I minimize the app, it's not shown in the list of existing windows. 
![image](https://user-images.githubusercontent.com/23486558/47835175-edb34c00-dde5-11e8-939f-ab42461d6709.png)
Clicking "Show All Windows" on the icon shows that there are no available windows, but the minimized window appears in the bar for recent files. 
![image](https://user-images.githubusercontent.com/23486558/47835396-2a337780-dde7-11e8-88f6-89ae968dbf11.png)

**Comment by almarklein** [comment]
Thanks for reporting this!

This looks similar: https://stackoverflow.com/questions/35591319/restore-minimized-app-osx-dock-click-event-pyqt-pyinstaller

Is this the same as a "system tray icon"? If so, this might also be relevant: https://doc.qt.io/qt-5.11/qtwidgets-desktop-systray-example.html

**Comment by balayanr** [comment]
The first link does talk about a similar problem (when the "minimization into icon" is turned off, it creates document-looking windows in the dock when you minimize).

The system tray solution is a bit strange, wondering if other QT-based apps have this problem. Keeping a system tray icon for something that doesn't really need to be there doesn't sound like a good idea, especially since you can't hide those on Mac.

**Comment by almarklein** [comment]
> Keeping a system tray icon for something that doesn't really need to be there doesn't sound like a good idea

I agree. I thought that it was maybe related. Good to hear that its not.

Any idea for a solution? I cannot imagine Pyzo being the first Qt app encountering this problem.

**Comment by balayanr** [comment]
I am not familiar with Qt, but from some googling it looks like for Mac "hide" and "minimize" mean different things (the former literally hides the window, the latter actually minimizes it, https://forum.qt.io/topic/10204/minimize-application-to-dock-icon, https://stackoverflow.com/questions/44286102/minimize-to-tray-on-macosx). 
I am not exactly sure how it's implemented in pyzo, but that would be a thing to check. 
One a solution would be to list the window and/or open files in the app menu in the dock, there is a platform-specific function handling that (http://doc.qt.io/archives/qt-4.8/exportedfunctions.html). Then, when a file or the icon is clicked, forcefully "unminimize" the window. (https://stackoverflow.com/questions/15143369/qt-on-os-x-how-to-detect-clicking-the-app-dock-icon)

I would look into the documentation on window-related functions, maybe there's a note mentioning unusual behaviour on MacOS.

**Comment by almarklein** [comment]
To get an idea of the urgency of this issue, what are the exact implications? Are you able to re-open Pyzo after it's minimized?

**Comment by balayanr** [comment]
Until I realized that the window appears in the list when clicking "Show all windows" I was sure that pyzo just broke every time I minimized it, because clicking on the icon didn't do anything. For years I just assumed people know about this issue and it was gonna get fixed eventually, but I guess there are not that many people with this system setting turned on. Or nobody bothered to report it, just like me.

Now that I know that there is a way to find the window it's not so bad. If the fix is really tough I wouldn't put it on top of priorities, but I think capturing the icon click and forcefully un-maximizing the window should do the trick. 
Listing the open files/windows in the dock menu (right click on the icon, like on the screenshots below) would be a nice addition, so that there was a way to un-minimize and have the needed file open, but it's fine either way. 
![image](https://user-images.githubusercontent.com/23486558/48073237-aed82880-e221-11e8-9ebd-b3cff81ceaff.png)
![image](https://user-images.githubusercontent.com/23486558/48073329-dd560380-e221-11e8-803e-76656b1d1477.png)

Thanks for taking a look at this issue

**Comment by almarklein** [comment]
I asked because this would need some extra work, involving me getting my hands on a mac :)  I will have a look sometime soon.

**Comment by bdieterm** [comment]
If the problem still exists after more than 7 years, please create a new issue.
